File: algorithms/Data/Stack/infixToPostfix.js
"use strict"
class Stack {
    constructor() {
        this.items = new Array()
    }

    isEmpty() {
        return this.items.length == 0
    }

    push(item) {
        return this.items.push(item)
    }

    pop() {
        return this.items.pop()
    }

    peek() {
        return this.items[this.items.length - 1]
    }

    size() {
        return this.items.length
    }
}

function infixToPostfix(infixexpr) {
   let precedence = {
        '*': 3,
        '/': 3,
        '+': 2,
        '-': 2,
        '(': 1
   }
   let opStack = new Stack()
   let postfix = []
   let tokenList = infixexpr.split(" ")

   tokenList.forEach(function(token, index) {
        if ("ABCDEFGHIJKLMNOPQRSTUVWXYZ".indexOf(token) != -1 || "0123456789".indexOf(token) != -1) {
            postfix.push(token)
        } else if (token == '(') {
            opStack.push(token)
        } else if (token == ')') {
            let topToken = opStack.pop()
            while(topToken != "(") {
                postfix.push(topToken)
                topToken = opStack.pop()
            }
        } else {
            while(!opStack.isEmpty() && (precedence[opStack.peek()] >= precedence[token])) {
                postfix.push(opStack.pop())
            }
            opStack.push(token)
        }
   });

   while(!opStack.isEmpty()) {
    postfix.push(opStack.pop())
   }

   return postfix.join(" ")
}

console.log(infixToPostfix("A + ( B * C )"));

File: algorithms/Data/Stack/infixToPostfix.py
from stack import Stack

def infixToPostfix(infixexpr):
    precedence = {}
    precedence['*'] = 3
    precedence['/'] = 3
    precedence['+'] = 2
    precedence['-'] = 2
    precedence['('] = 1
    opstack = Stack()
    postfixLIst = [] # output list
    tokenList = infixexpr.split()

    for token in tokenList:
        if token in "ABCDEFGHIJKLMNOPQRSTUVWXYZ" or token in "0123456789":
            postfixLIst.append(token)
        elif token == '(':
            opstack.push(token)
        elif token == ')':
            topToken = opstack.pop()
            while topToken != '(':
                postfixLIst.append(topToken)
                topToken = opstack.pop()
        else:
            while (not opstack.isEmpty()) and (precedence[opstack.peek()] >= precedence[token]):
                postfixLIst.append(opstack.pop())
            opstack.push(token)

    while not opstack.isEmpty():
        postfixLIst.append(opstack.pop())

    return " ".join(postfixLIst)

print(infixToPostfix("A + ( B * C )"))

